include/logger.h

```diff
@@ -0,0 +1,175 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef SMARTREDIS_LOGGER_H
+#define SMARTREDIS_LOGGER_H
+
+#include <stdlib.h>
+#include <string>
+#include "utility.h"
+#include "sr_enums.h"
+
+///@file
+
+namespace SmartRedis {
+
+/*!
+*   \brief The Logger class implements a logging facility for SmartRedis
+*/
+class Logger {
+
+    public:
+
+        /*!
+        *   \brief Retrieve the unique Logger instance
+        *   \returns The actual logger instance
+        */
+       static Logger& get_instance()
+       {
+           static Logger instance; // instantiated on first acccess
+           return instance;
+       }
+
+        /*!
+        *   \brief Set up logging for the current client
+        *   \param _client_id ID to use for the current client
+        */
+       void configure_logging(const std::string& _client_id);
+
+    private:
+
+        /*!
+        *   \brief Logger constructor
+        */
+        Logger() { initialized = false; }
+
+        /*!
+        *   \brief Logger copy constructor unavailable
+        *   \param logger The Logger to copy for construction
+        */
+        Logger(const Logger& logger) = delete;
+
+        /*!
+        *   \brief Logger assignment operator unavailable
+        *   \param logger The Logger to copy for construction
+        */
+        void operator=(const Logger& logger) = delete;
+
+        /*!
+        *   \brief Default Logger destructor
+        */
+        ~Logger() = default;
+
+    public:
+
+        /*!
+        *   \brief Conditionally log data if the logging level is high enough
+        *   \param level Minimum logging level for data to be logged
+        *   \param data Text of data to be logged
+        */
+        void log_data(SRLoggingLevel level, const char* data);
+
+        /*!
+        *   \brief Conditionally log data if the logging level is high enough
+        *   \param level Minimum logging level for data to be logged
+        *   \param data Text of data to be logged
+        */
+        void log_data(SRLoggingLevel level, const std::string& data);
+
+        /*!
+        *   \brief Conditionally log data if the logging level is high enough
+        *   \param level Minimum logging level for data to be logged
+        *   \param data Text of data to be logged
+        */
+        void log_data(SRLoggingLevel level, const std::string_view& data);
+
+    private:
+
+        /*!
+        *   \brief Track whether logging is initialized
+        */
+        bool initialized;
```

Comment:
In other files I think we have been using ``_`` before private member variables.  It might be nice to stay consistent with other conventions.  This is a general comment and won't comment all new classes to avoid redundant comments.

Reply:
Done, thanks for noticing that I got the convention backwards here